### PR TITLE
Strategic AI: cooperating heroes and improved estimates

### PR DESCRIPTION
**Pull request by idshibanov**
Now we can evaluate and compare army reinforcements, castle buildings, heroes, skills, spells and artifacts to make more accurate strategic decisions in the future.

Taught AI heroes to exchange armies and artifacts.
Added a few additional validation checks to avoid heroes getting stuck on each other.
Made few improvements based on newly implemented tools (castle/hero priority for example).

### CODE REVIEW

File: src/fheroes2/ai/normal/ai_normal_hero.cpp [hunk 1]
```diff
@@ -29,34 +29,57 @@
...
+            else
+                return castle->getBuildingValue() * 70.0 + 1250;
         }
         else if ( objectID == MP2::OBJ_HEROES ) {
+            const Heroes * otherHero = tile.GetHeroes();
+            if ( !otherHero )
+                return valueToIgnore;
+
+            if ( hero.GetColor() == otherHero->GetColor() ) {
+                if ( hero.getStatsValue() >= otherHero->getStatsValue() + 1 )
```
Comment (ihhub) [inline src/fheroes2/ai/normal/ai_normal_hero.cpp:54]: Why did we put +1?

File: src/fheroes2/army/army.cpp [hunk 2]
```diff
@@ -372,6 +372,15 @@ u32 Troops::GetUniqueCount( void ) const
     return monsters.size();
 }
 
+double Troops::GetStrength() const
+{
+    double strength = 0;
+    for ( const Troop * troop : *this )
+        if ( troop && troop->isValid() )
+            strength += troop->GetStrength();
```
Comment (ihhub) [inline src/fheroes2/army/army.cpp:380]: Please change it into:
```
for ( const Troop * troop : *this ) {
    if ( troop && troop->isValid() )
        strength += troop->GetStrength();
}
```

so it'll be easier to separate on loop from another :)

File: src/fheroes2/castle/castle.cpp [hunk 3]
```diff
@@ -397,6 +397,50 @@ void Castle::EducateHeroes( void )
...
+    if ( isBuild( BUILD_CASTLE ) )
+        value += 5;
+
+    if ( isBuild( DWELLING_MONSTER6 ) )
+        value += 6;
+
+    // DWELLING_UPGRADE7 resolves to a negative, can't use <= operator
+    for ( uint32_t upgrade = DWELLING_UPGRADE2; upgrade != DWELLING_UPGRADE7; upgrade <<= 1 )
+        if ( isBuild( upgrade ) )
+            ++value;
```
Comment (ihhub) [inline src/fheroes2/castle/castle.cpp:414]: Please add {} to the first loop to make the code little more easier to read.

Comment (ihhub) [reply src/fheroes2/castle/castle.cpp:414]: Shall we have a special case only for `DWELLING_UPGRADE7` with Warlock castle?

File: src/fheroes2/castle/castle.cpp [hunk 4]
```diff
@@ -397,6 +397,50 @@ void Castle::EducateHeroes( void )
...
+    double spellValue = 0;
+    const SpellStorage & guildSpells = mageguild.GetSpells( GetLevelMageGuild(), isLibraryBuild() );
+    for ( const Spell & spell : guildSpells ) {
+        if ( hero.CanLearnSpell( spell ) && !hero.HaveSpell( spell, true ) ) {
+            spellValue += spell.Level() * 250.0;
+        }
+    }
+
+    Troops reinforcement;
+    for ( u32 dw = DWELLING_MONSTER6; dw >= DWELLING_MONSTER1; dw >>= 1 )
```
Comment (ihhub) [inline src/fheroes2/castle/castle.cpp:437]: Please use `uint32_t` instead of `u32`.

File: src/fheroes2/heroes/skill.cpp [hunk 5]
```diff
@@ -638,6 +638,17 @@ int Skill::SecSkills::Count( void ) const
     return std::count_if( begin(), end(), std::mem_fun_ref( &Secondary::isValid ) );
 }
 
+int Skill::SecSkills::GetTotalLevel() const
+{
+    int result = 0;
+    for ( const Skill::Secondary skill : *this ) {
```
Comment (ihhub) [inline src/fheroes2/heroes/skill.cpp:644]: Please use const reference here.

File: src/fheroes2/resource/artifact.cpp [hunk 6]
```diff
@@ -812,6 +812,44 @@ u32 BagArtifacts::CountArtifacts( void ) const
     return static_cast<uint32_t>( std::count_if( begin(), end(), std::mem_fun_ref( &Artifact::isValid ) ) ); // no way that we have more than 4 billion artifacts
 }
 
+int BagArtifacts::getArtifactValue() const
+{
+    int result = 0;
+    for ( const Artifact & art : *this )
+        if ( art.isValid() )
+            result += art.getArtifactValue();
```
Comment (ihhub) [inline src/fheroes2/resource/artifact.cpp:820]: Please add {} to the first loop to make the code little more easier to read.

File: src/fheroes2/resource/artifact.cpp [hunk 7]
```diff
@@ -812,6 +812,44 @@ u32 BagArtifacts::CountArtifacts( void ) const
...
+}
+
+void BagArtifacts::exchangeArtifacts( BagArtifacts & giftBag )
+{
+    std::vector<Artifact> combined;
+    for ( auto it = begin(); it != end(); ++it )
+        if ( it->isValid() && it->GetID() != Artifact::MAGIC_BOOK ) {
+            combined.push_back( *it );
+            it->Reset();
+        }
```
Comment (ihhub) [inline src/fheroes2/resource/artifact.cpp:832]: Please add {} to the first loop to make the code little more easier to read.

File: src/fheroes2/resource/artifact.cpp [hunk 8]
```diff
@@ -812,6 +812,44 @@ u32 BagArtifacts::CountArtifacts( void ) const
...
+
+    for ( auto it = giftBag.begin(); it != giftBag.end(); ++it )
+        if ( it->isValid() && it->GetID() != Artifact::MAGIC_BOOK ) {
+            combined.push_back( *it );
+            it->Reset();
+        }
```
Comment (ihhub) [inline src/fheroes2/resource/artifact.cpp:838]: Please add {} to the first loop to make the code little more easier to read.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @idshibanov , I left multiple comments in this pull request. Could you please take a look into them? My worry that we need to separate Black Dragons upgrade from others.

Also I'm thinking that we might need to use `std::bitset` to hold many other items...

**Comment by idshibanov** [comment]
> 
> 
> Hi @idshibanov , I left multiple comments in this pull request. Could you please take a look into them? My worry that we need to separate Black Dragons upgrade from others.
> 
> Also I'm thinking that we might need to use `std::bitset` to hold many other items...

I agree that castle code should be changed to use bitset for buildings. We already ran out of bits without even adding new buildings. We can do this change after the release as castle code should be refactored too.

I'll move black dragons to a separate IF.

File: src/fheroes2/ai/normal/ai_normal_hero.cpp (hunk 1 above)
Comment (idshibanov) [reply src/fheroes2/ai/normal/ai_normal_hero.cpp:54]: Should have 2 level difference before offering troops to other hero.

Comment (ihhub) [reply src/fheroes2/ai/normal/ai_normal_hero.cpp:62]: Shouldn't this be `if ( hero.getStatsValue() > otherHero->getStatsValue() + 1 )` (without =)?

Comment (idshibanov) [reply src/fheroes2/ai/normal/ai_normal_hero.cpp]: Updated to +2 without =

File: src/fheroes2/castle/castle.cpp [hunk 9]
```diff
@@ -2219,14 +2268,24 @@ std::string Castle::String( void ) const
...
+    os << "name and type   : " << name << " (" << Race::String( race ) << ")" << std::endl
        << "color           : " << Color::String( GetColor() ) << std::endl
-       << "build           : "
-       << "0x" << std::hex << building << std::dec << std::endl
-       << "present boat    : " << ( PresentBoat() ? "yes" : "no" ) << std::endl
-       << "nearly sea      : " << ( HaveNearlySea() ? "yes" : "no" ) << std::endl
-       << "is castle       : " << ( isCastle() ? "yes" : "no" ) << std::endl
+       << "dwellings       : ";
+
+    for ( uint32_t level = 0; level < 6; ++level ) {
```
Comment (ihhub) [inline src/fheroes2/castle/castle.cpp:2275]: Here we miss DWELLING_UPGRADE7.

Comment (idshibanov) [reply src/fheroes2/castle/castle.cpp:2275]: It is a special case at this time, I don't think it's important to show here. If 6th dwelling is upgraded in Warlock castle we can safely assume it's up to Black dragons.

Comment (idshibanov) [reply src/fheroes2/castle/castle.cpp]: Cleaned it up a little bit

File: src/fheroes2/kingdom/kingdom.cpp [hunk 10]
```diff
@@ -379,7 +379,7 @@ bool Kingdom::isValidKingdomObject( const Maps::Tiles & tile, int objectID ) con
     // Hero object can overlay other objects when standing on top of it: force check with GetObject( true )
     if ( objectID == MP2::OBJ_HEROES ) {
         const Heroes * hero = tile.GetHeroes();
-        return hero && !Players::isFriends( color, hero->GetColor() );
+        return hero && ( color == hero->GetColor() || !Players::isFriends( color, hero->GetColor() ) );
```
Comment (ihhub) [inline src/fheroes2/kingdom/kingdom.cpp:382]: How is it possible to walk through of existing hero?

**Comment by ihhub** [comment]
@idshibanov excellent work!

**Pull request merged**